addons/pos_loyalty/static/src/overrides/models/pos_store.js

```diff
@@ -341,25 +330,81 @@ patch(PosStore.prototype, {
      * @param {int} partnerId
      */
     async fetchLoyaltyCard(programId, partnerId) {
-        for (const coupon of Object.values(this.couponCache)) {
-            if (coupon.partner_id === partnerId && coupon.program_id === programId) {
-                return coupon;
-            }
+        const coupon = this.models["loyalty.card"].find(
+            (c) => c.partner_id?.id === partnerId && c.program_id?.id === programId
+        );
+        if (coupon) {
+            return coupon;
         }
         const fetchedCoupons = await this.fetchCoupons([
             ["partner_id", "=", partnerId],
             ["program_id", "=", programId],
         ]);
         const dbCoupon = fetchedCoupons.length > 0 ? fetchedCoupons[0] : null;
-        return dbCoupon || new PosLoyaltyCard(null, null, programId, partnerId, 0);
+        return (
+            dbCoupon ||
+            this.models["loyalty.card"].create({
+                id: loyaltyIdsGenerator(),
+                code: null,
+                program_id: this.models["loyalty.program"].get(programId),
+                partner_id: this.models["res.partner"].get(partnerId),
+                points: 0,
+                // TODO JCB: expiration_date
+            })
+        );
     },
     getLoyaltyCards(partner) {
         const loyaltyCards = [];
         if (this.partnerId2CouponIds[partner.id]) {
             this.partnerId2CouponIds[partner.id].forEach((couponId) =>
-                loyaltyCards.push(this.couponCache[couponId])
+                loyaltyCards.push(this.models["loyalty.card"].get(couponId))
             );
         }
         return loyaltyCards;
     },
+    /**
+     * IMPROVEMENT: It would be better to update the local order object instead of creating a new one.
+     *   - This way, we don't need to remember the lines linked to negative coupon ids and relink them after pushing the order.
+     */
+    async push_single_order(...args) {
+        const [orderToPush] = args;
+
+        // Keep some linked records to relink them after pushing the order to the server.
+        const lineCouponMap = orderToPush.lines.reduce((agg, line) => {
+            if (line.coupon_id && line.coupon_id.id < 0) {
+                return { ...agg, [line.uuid]: line.coupon_id.id };
+            } else {
+                return agg;
+            }
+        }, {});
+        const lineRewardProductMap = orderToPush.lines.reduce((agg, line) => {
+            if (line.reward_product_id) {
+                return { ...agg, [line.uuid]: line.reward_product_id.id };
+            } else {
+                return agg;
+            }
+        }, {});
+
+        const [updatedOrder] = await super.push_single_order(...args);
```

Comment:
This methods isn't awaitable, the variable `updatedOrder` do not contains the result. What you can do is call directly `syncAllOrders` which return serverOrders.

If you take this option don't forget to put your order in pendingOrder before call syncAllOrders. (`addPendingOrder` method)

Reply:
Ah it seems to return a promise from the mutex, but can you check anyway ?